### ISSUE TITLE
Search Data Store key for particular value [ACTION]

### ISSUE THREAD

**Issue by NebularNerd** [PipedreamHQ/pipedream#5223]
This Data Store lacks a feature to check for the existence of a record within a key. I can do this inside Python but I wonder if a dedicated action would be faster/lighter.

Say for example we have a few values stored in a data store and I want to find a match (say 12345):
```
[
  67890,
  12345,
  7063578509386
]
```
I've tried every Data Store action relating to reading records but they dump back the whole JSON, Pipedream Filter is unable to search the records so cannot provide a match to stop execution with. so I guess an action such as  
`Search key for matching value`  
with a TRUE/FALSE return should do the trick. 

Any help on this would be great, I have no issue with the Python method but this seems a fairly easy win for those just wanting to avoid coding.

**Comment by vunguyenhung** [comment]
Hi @NebularNerd, I think it makes sense to add this action. One thing is user need to input the data type of the value, so that Pipedream know how to check

**Comment by NebularNerd** [comment]
> Hi @NebularNerd, I think it makes sense to add this action. One thing is user need to input the data type of the value, so that Pipedream know how to check

Yes this would make sense, as the values could be strings or raw array data.

**Comment by andrewjschuang** [comment]
Hi, @NebularNerd, the Filter action has simple `array` and `object` searching conditions, and the below worked for me:

![image](https://user-images.githubusercontent.com/15385076/213687741-6c0397d2-00a7-44f6-8820-05ef29f0d2c6.png)

![image](https://user-images.githubusercontent.com/15385076/213687297-e064b601-a875-4e11-9a18-9b19a7d24c63.png)

**Comment by NebularNerd** [comment]
@andrewjschuang Awesome, that is exactly what I was looking for. I got overly focused on the idea of the Data Store being able to search for the key I did not look more into the filters full power. This should reduce my Python runtime drastically in the workflow.

@vunguyenhung  @dylburger This provides me an immediate solution, shall I close this issue or would you like to leave it open until the new Data Store action is added?

**Comment by andrewjschuang** [comment]
\+ @dannyroosevelt

**Comment by dannyroosevelt** [comment]
Interesting idea. So you'd search through the data store records based on the record's value instead of key, and return the key?

**Comment by NebularNerd** [comment]
> Interesting idea. So you'd search through the data store records based on the record's value instead of key, and return the key?

Kind of, I was thinking more along the lines of a True/False return if the record exists, using my super lazy example you have a Data Store JSON with 
```
{
	"subject_id": [
		67890,
		12345,
		7063578509386
	]
}
```
You run a search for `12345` (in real life this would normally be a value in the trigger data or a previous step) and as it exists you get a return of TRUE, maybe an additional return with the JSON path to that record would be helpful as well for other purposes.

This is kind of similar to @andrewjschuang suggestion that is present in the filter, however, the filter works but it's only a stop/go switch (which works for me at present). This new action would allow for finer control for example creating a whitelist or similar, by checking if a specific record exists you could then have some sort of logic in the next step, so for Python you might have:
```
if [pd.step][search.for.record] = true:
    print ('We found the record')
    # Do stuff
else:
   print('Record not found')
   # Do other stuff
```

I have a very similar logic coded in Python using json.dumps() and other kludges to run a stop/go filter at the next step and merge a new record into the data store, but this would be a smoother and hopefully more lightweight solution. 

**Potential Issues:** Nesting levels, duplicate records under different levels, string OR int and anything else JSON related. Duplicates are not a huge issue, Data Store path selection logic should avoid that.

**Comment by vellames** [comment]
That's feasible. Probably will not perform very well in a large DataStore, but feasible :)

**Comment by NebularNerd** [comment]
> That's feasible. Probably will not perform very well in a large DataStore, but feasible :)

I imagine the performance would be similar to the Filter's ARRAY search speeds, from @andrewjschuang solution above. I agree with a huge database it may take some time which could be an issue with hitting Pipedream limits.

For my purposes I am running the check to see if the ID of a product has been logged before to prevent the rest of the script running, sensibly I could trim/delete the database every few months, it's mainly to stop the same ID being processed repeatedly as the trigger I use generates duplicate events (due to the way the source generates the trigger events, not an issue)

**Comment by Hassan-Abbas-SQA** [comment]
@vellames TESTFAILED

1. can you please add a condition that it only accepts the valid numerical or character
![image](https://user-images.githubusercontent.com/104946677/214637631-a4813b72-729f-4a8c-b859-2bcc4b4c27dd.png)
2: its not working if i entered half value of the original to find the key like original value is 12345 i entered 123 it returned nothing
<img width="896" alt="image" src="https://user-images.githubusercontent.com/104946677/214638206-710f5ce2-e634-403f-b281-47e2eb7ec0db.png">

**Comment by NebularNerd** [comment]
@Hassan-Abbas-SQA some interesting tests, there's some things I had actually not considered. Partial or exact matching is not something I considered based on my use case, the same for case sensitivity. Maybe these could be optional settings?

**Comment by Hassan-Abbas-SQA** [comment]
> @Hassan-Abbas-SQA some interesting tests, there's some things I had actually not considered. Partial or exact matching is not something I considered based on my use case, the same for case sensitivity. Maybe these could be optional settings?

yes @vellames we can create an optional boolean prop for partial match or exact match to give the user complete power over search

Initially i did these tests so that we can modify the prop description to guide the user so he doesn’t get confused

**Comment by vellames** [comment]
I did it to return an exactly match, I can add these conditions

**Comment by Hassan-Abbas-SQA** [comment]
Thanks @vellames

**Comment by vellames** [comment]
@Hassan-Abbas-SQA can you check it please?

**Comment by Hassan-Abbas-SQA** [comment]
@vellames OOPS the action returned the key correct but the key starts with query itself which is making the key invalid
<img width="820" alt="image" src="https://user-images.githubusercontent.com/104946677/215208096-f58da003-e3e7-421f-8b56-4b75a51b3da8.png">
<img width="820" alt="image" src="https://user-images.githubusercontent.com/104946677/215208934-03019a98-a4a8-4cc8-8253-c839ca0fcb1a.png">
other than that tested possible ways and it work really great @vellames please fix the only remaining issue then we are good to go 🚀

**Comment by vellames** [comment]
Oops, it should be fine now

**Comment by Hassan-Abbas-SQA** [comment]
### ALL-TESTSPASSED
Great Stuff @vellames 
<img width="606" alt="image" src="https://user-images.githubusercontent.com/104946677/215828993-00f302e8-91bb-43b4-a9e1-34de674c69f6.png">

**Comment by NebularNerd** [comment]
Sorry for not doing so earlier but thank you @vellames, @Hassan-Abbas-SQA and everyone else for making this happen. I've just found time to have a go of this and it works exactly how I envisioned it. 🙂

I think with some refactoring on my workflow I was looking to use it on, this will cut down on operation time by not having to go through so many filter steps. I can also test out my new Python-fu skills I've been teaching myself to further cut down excess work being done.
